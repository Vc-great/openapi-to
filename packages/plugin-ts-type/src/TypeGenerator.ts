import path from "node:path";

import { UUID_TAG_NAME } from "@openapi-to/core/utils";

import _ from "lodash";

import { Component } from "./Component.ts";
import {
  TYPE_MODEL_FOLDER_NAME,
  TYPE_SUFFIX,
  UUID_PREFIX,
} from "./constants.ts";
import { EnumGenerator } from "./EnumGenerator.ts";
import { Schema } from "./Schema.ts";

import type { Operation } from "oas/operation";
import type { SchemaObject } from "oas/types";
import type {
  ImportDeclarationStructure,
  InterfaceDeclarationStructure,
  ModuleDeclarationStructure,
  TypeAliasDeclarationStructure,
} from "ts-morph";
import type { TypeOldNode } from "./TypeOldNode.ts";
import type { Config } from "./types.ts";
type TypeStatements =
  | InterfaceDeclarationStructure
  | TypeAliasDeclarationStructure;

type ImportStatementsOmitKind = Omit<ImportDeclarationStructure, "kind">;

type ResponseObject = {
  code: string;
  jsonSchema?: {
    description?: string;
    label: string;
    schema: SchemaObject;
    type: string | string[];
  };
};

export class TypeGenerator {
  private operation: Operation | undefined;
  private oas: Config["oas"];
  private readonly openapi: Config["openapi"];
  private readonly ast: Config["ast"];
  private readonly pluginConfig: Config["pluginConfig"];
  private readonly openapiToSingleConfig: Config["openapiToSingleConfig"];
  private enumGenerator: EnumGenerator;
  private importCache: Set<string> = new Set<string>();
  private component: Component;
  private schema: Schema;
  private readonly modelFolderName: string = TYPE_MODEL_FOLDER_NAME;
  private oldNode: TypeOldNode;
  constructor(config: Config) {
    this.oas = config.oas;
    this.ast = config.ast;
    this.pluginConfig = config.pluginConfig;
    this.openapiToSingleConfig = config.openapiToSingleConfig;
    this.openapi = config.openapi;

    this.schema = new Schema(config);
    this.oldNode = config.oldNode;
    this.component = new Component(config);
    this.enumGenerator = EnumGenerator.getInstance(config);
  }

  get compare(): boolean {
    return this.pluginConfig?.compare || false;
  }

  get namespaceUUID(): string {
    return UUID_PREFIX + this.openapi.currentTagNameOfPinYin;
  }
  get nameSpaceName(): string {
    return _.upperFirst(this.openapi.currentTagNameOfPinYin);
  }

  get fileName(): string {
    return (
      this.oldNode.baseName ||
      _.lowerFirst(this.openapi.currentTagNameOfPinYin) + "." + TYPE_SUFFIX
    );
  }

  build(): void {
    _.forEach(this.openapi.pathGroupByTag, (pathGroup, tag) => {
      this.oldNode.setCurrentSourceFile(UUID_PREFIX + _.camelCase(tag));
      this.enumGenerator.setPrefix(this.nameSpaceName);
      const typeStatements = _.chain(pathGroup)
        .map(({ path, method, tag }) => {
          this.operation = this.openapi.setCurrentOperation(path, method, tag);

          return _.chain([] as Array<TypeStatements | null>)
            .concat(this.generateParametersType())
            .concat(this.generateRequestBodyType())
            .concat(this.generateCodeByResponse())
            .filter(Boolean)
            .value();
        })
        .flatten()
        .filter((x) => !_.isEmpty(x))
        .value();

      const filePath = path.resolve(
        this.openapiToSingleConfig.output.dir,
        this.fileName + ".ts",
      );

      const refKey = [...this.openapi.refCache.keys()];

      this.ast.createSourceFile(filePath, {
        statements: [
          ...this.generateImport(refKey),
          ...this.enumGenerator.generateEnum(),
          this.generateNameSpace(typeStatements),
        ],
      });

      this.openapi.resetRefCache();
    });

    this.enumGenerator.resetPrefix();
    this.component.generateCodeByComponent();

    this.component.generateModelIndex();
  }

  /**
   *
   * @example
   * ```
   * // ts type
   * import type { DrmsDynamicDataType } from './drmsDynamicDataZod'
   * ```
   */
  generateImport(
    importModel: Array<string>,
  ): Array<ImportDeclarationStructure> {
    const model = _.chain(importModel)
      .map(($ref: string) => {
        const name = _.upperFirst(this.openapi.getRefAlias($ref));
        const UUID = UUID_PREFIX + name;
        const declaration = this.oldNode.declarationCache.get(UUID);
        return declaration?.getName() ?? name;
      })
      .value();

    const typeModel: ImportStatementsOmitKind = {
      namedImports: [...model],
      isTypeOnly: true,
      moduleSpecifier: "./" + this.modelFolderName,
    };

    const statements = _.chain([] as Array<ImportStatementsOmitKind>)
      .concat(_.isEmpty(model) ? [] : typeModel)
      .filter(Boolean)
      .value();

    return this.ast.generateImportStatements(statements);
  }

  /**
   * @example
   * ```
   * export namespace Tag {
   *  export type queryRequest = object
   * }
   * ```
   */
  generateNameSpace(
    typeStatements: Array<TypeStatements>,
  ): ModuleDeclarationStructure {
    return this.ast.generateModuleStatements({
      docs: [
        {
          tags: [
            {
              tagName: "tag",
              text: this.openapi.currentTagName,
            },
            {
              tagName: "description",
              text: this.openapi.currentTagMetadata?.description,
            },
            ...(this.pluginConfig?.compare
              ? [
                  {
                    tagName: UUID_TAG_NAME,
                    text: this.namespaceUUID,
                  },
                ]
              : []),
          ].filter((x) => x.text),
        },
      ],
      isExported: true,
      name: this.oldNode.namespaceName ?? this.nameSpaceName,
      statements: typeStatements,
    });
  }

  generateParametersType(): Array<
    InterfaceDeclarationStructure | TypeAliasDeclarationStructure
  > {
    if (this.openapi.upperFirstRequestName === "FindByStatusGet") {
      //debugger;
    }
    const queryParamsName = this.openapi.upperFirstRequestName + "QueryParams";
    const queryStatements = this.ast.generateInterfaceStatements({
      isExported: true,
      name: queryParamsName,
      docs: [
        {
          tags: [
            {
              leadingTrivia: "\n",
              tagName: "description",
              text: "queryParams",
            },
          ],
        },
      ],
      properties:
        this.schema.getBaseTypeFromSchema(
          this.openapi.parameter?.getParametersSchema("query") || null,
          this.nameSpaceName + queryParamsName,
        ) || [],
    });
    const pathParamsName = this.openapi.upperFirstRequestName + "PathParams";
    const pathStatements = this.ast.generateInterfaceStatements({
      isExported: true,
      name: pathParamsName,
      docs: [
        {
          tags: [
            {
              leadingTrivia: "\n",
              tagName: "description",
              text: "pathParams",
            },
          ],
        },
      ],
      properties:
        this.schema.getBaseTypeFromSchema(
          this.openapi.parameter?.getParametersSchema("path") || null,
          this.nameSpaceName + pathParamsName,
        ) || [],
    });
    //[mediaType, bodySchema, description]

    return _.chain(
      [] as Array<
        InterfaceDeclarationStructure | TypeAliasDeclarationStructure
      >,
    )
      .concat(this.openapi.parameter?.hasQueryParameters ? queryStatements : [])
      .concat(this.openapi.parameter?.hasPathParameters ? pathStatements : [])
      .filter(Boolean)
      .value();
  }

  generateCodeByResponse(): Array<
    InterfaceDeclarationStructure | TypeAliasDeclarationStructure
  > {
    const codes = this.openapi.response?.getResponseStatusCodes || [];

    const successCode = (codes || []).filter((code) =>
      /^(2[0-9][0-9]|300)$/.test(code),
    );
    const errorCode = (codes || []).filter((code) =>
      /^([3-5][0-9][0-9])$/.test(code),
    );
    const isSuccessCode = (code: string) => /^(2[0-9][0-9]|300)$/.test(code);

    const responseObject = _.chain([...successCode, ...errorCode])
      .map((code) => {
        return {
          code,
          jsonSchema:
            _.head(this.openapi.response?.getResponseAsJSONSchema(code)) ||
            null,
        };
      })
      .filter((x) => !_.isNull(x.jsonSchema))
      .value() as Array<ResponseObject>;

    const errorType = this.ast.generateTypeAliasStatements({
      name: this.openapi.upperFirstRequestName + "Error",
      type:
        _.chain(errorCode)
          .filter((code) => responseObject.map((x) => x.code).includes(code))
          .map((code) => this.openapi.upperFirstResponseName + code)
          .join("|")
          .value() || "unknown",
      //docs: [{ description: "" }],
      isExported: true,
    });

    const successDefaultType = this.ast.generateTypeAliasStatements({
      name: this.openapi.upperFirstResponseName,
      type: "unknown",
      docs: [], //[{ description: "" }],
      isExported: true,
    });

    return _.chain(responseObject)
      .map((responseObject) =>
        this.generateResponseSingleSchema(responseObject),
      )
      .concat(errorType)
      .concat(_.isEmpty(successCode) ? successDefaultType : [])
      .filter(Boolean)
      .value();
  }

  generateResponseSingleSchema({
    code,
    jsonSchema,
  }: ResponseObject):
    | InterfaceDeclarationStructure
    | TypeAliasDeclarationStructure {
    const schema = jsonSchema?.schema;
    const description = jsonSchema?.description;
    const isError = /^([3-5][0-9][0-9])$/.test(code);
    const name = this.openapi.upperFirstResponseName + (isError ? code : "");

    if (!schema) {
      return this.ast.generateTypeAliasStatements({
        name,
        type: "unknown",
        docs: [], //[{ description: "" }],
        isExported: true,
      });
    }

    if (this.openapi.isReference(schema)) {
      return this.ast.generateTypeAliasStatements({
        name,
        type: _.upperFirst(this.openapi.getRefAlias(schema.$ref)),
        docs: description
          ? [
              {
                tags: [
                  {
                    leadingTrivia: "\n",
                    tagName: "description",
                    text: description,
                  },
                ],
              },
            ]
          : [],
        isExported: true,
      });
    }

    if (schema.type === "object") {
      return this.ast.generateInterfaceStatements({
        isExported: true,
        name,
        docs: description
          ? [
              {
                tags: [
                  {
                    leadingTrivia: "\n",
                    tagName: "description",
                    text: description,
                  },
                ],
              },
            ]
          : [],
        properties: this.schema.getBaseTypeFromSchema(schema) || [],
      });
    }

    return this.ast.generateTypeAliasStatements({
      name,
      type: this.schema.formatterSchemaType(
        schema,
        this.nameSpaceName + _.upperFirst(jsonSchema?.label),
      ),
      docs: description
        ? [
            {
              tags: [
                {
                  leadingTrivia: "\n",
                  tagName: "description",
                  text: description,
                },
              ],
            },
          ]
        : [],
      isExported: true,
    });
  }

  //SchemaObject
  generateRequestBodyType(): Array<TypeStatements> | null {
    const _bodySchema = this.openapi.requestBody?.getRequestBodySchema();
    if (_bodySchema === undefined || _.isBoolean(_bodySchema)) {
      return null;
    }
    let schema: SchemaObject | null = null;
    if ("schema" in _bodySchema) {
      schema = _bodySchema.schema || (null as SchemaObject | null);
    }

    if (_.isArray(_bodySchema)) {
      schema = _.get(_bodySchema, "[1].schema", null) as SchemaObject | null;
    }

    if (schema === null) {
      return null;
    }

    if (this.openapi.isReference(schema)) {
      return [
        this.ast.generateTypeAliasStatements({
          name: this.openapi.upperFirstBodyDataName,
          type: _.upperFirst(this.openapi.getRefAlias(schema.$ref)),
          docs: [], //[{ description: "" }],
          isExported: true,
        }),
      ];
    }

    if (schema.type === "array") {
      return [
        this.ast.generateTypeAliasStatements({
          name: this.openapi.upperFirstBodyDataName,
          type: this.schema.formatterSchemaType(
            schema,
            this.nameSpaceName + "BodyData",
          ),
          docs: schema.description
            ? [
                {
                  tags: [
                    {
                      leadingTrivia: "\n",
                      tagName: "description",
                      text: schema.description,
                    },
                  ],
                },
              ]
            : [],
          isExported: true,
        }),
      ];
    }

    return [
      this.ast.generateInterfaceStatements({
        isExported: true,
        name: this.openapi.upperFirstBodyDataName,
        docs: schema.description
          ? [
              {
                tags: [
                  {
                    leadingTrivia: "\n",
                    tagName: "description",
                    text: schema.description,
                  },
                ],
              },
            ]
          : [],
        properties: this.schema.getBaseTypeFromSchema(schema) || [],
      }),
    ];
  }
}
