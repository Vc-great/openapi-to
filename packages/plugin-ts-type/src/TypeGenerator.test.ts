import path, { dirname } from "node:path";
import { fileURLToPath } from "node:url";

import { AST, OpenAPI } from "@openapi-to/core";

import _ from "lodash";
import Oas from "oas";

import petStore from "../mock/petstore.json";
import { TypeGenerator } from "./TypeGenerator.ts";
import { TypeOldNode } from "./TypeOldNode.ts";

import type { OpenapiToSingleConfig } from "@openapi-to/core";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

describe("typerGenerator", async () => {
  const openapiToSingleConfig: OpenapiToSingleConfig = {
    name: "",
    root: "",
    input: {
      path: "",
    },
    output: {
      dir: path.resolve(__dirname, "../mock"),
    },
    plugins: [],
  };

  test("TypeGenerator getFullText", () => {
    const pluginConfig = {
      compare: false,
    };
    const ast = new AST();
    // @ts-expect-error Not a canonical document
    const oas = new Oas(petStore);
    const openapi = new OpenAPI({}, oas);
    const oldNode = new TypeOldNode(pluginConfig, openapiToSingleConfig);
    const typerGenerator = new TypeGenerator({
      oas,
      ast,
      openapi,
      pluginConfig,
      openapiToSingleConfig,
      oldNode,
    });
    typerGenerator.build();
    const text = _.chain(ast.sourceFile)
      .map((sourceFile) => sourceFile.getFullText())
      .join("\n")
      .value();

    expect(text).toMatchSnapshot();
  });

  test("compare TypeGenerator getFullText", () => {
    const pluginConfig = {
      compare: true,
    };
    const ast = new AST();
    // @ts-expect-error Not a canonical document
    const oas = new Oas(petStore);
    const openapi = new OpenAPI({}, oas);
    const oldNode = new TypeOldNode(pluginConfig, openapiToSingleConfig);
    const typerGenerator = new TypeGenerator({
      oas,
      ast,
      openapi,
      pluginConfig,
      openapiToSingleConfig,
      oldNode,
    });
    typerGenerator.build();
    const text = _.chain(ast.sourceFile)
      .map((sourceFile) => sourceFile.getFullText())
      .join("\n")
      .value();

    expect(text).toMatchSnapshot();
  });
});
