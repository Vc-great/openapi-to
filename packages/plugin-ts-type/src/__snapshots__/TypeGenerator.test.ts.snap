// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`typerGenerator > TypeGenerator getFullText 1`] = `
"import type { TestDto2, Test32145, ApiResponse, Pet } from "./typeModels";

/**
 * @tag pet
 * @description Everything about your Pets
 * @UUID type-pet
 */
export namespace Pet {
    /** OK */
    export type TestPostMutationResponse = TestDto2;
    export type TestPostError = unknown;
    /** OK */
    export type TestPutMutationResponse = TestDto2;
    export type TestPutError = unknown;
    export type DelByTestMutationRequest = Array<number>;
    /** OK */
    export type DelByTestMutationResponse = Test32145;
    export type DelByTestError = unknown;

    /** queryParams */
    export interface TestIdGetQueryParams {
        fields?: Array<string>;
        page: number;
        size: number;
    }

    /** pathParams */
    export interface TestIdGetPathParams {
        testId?: number;
        testId2?: string;
    }

    /** OK */
    export type TestIdGetQueryResponse = TestDto2;
    export type TestIdGetError = unknown;

    /** pathParams */
    export interface UploadImagePostPathParams {
        petId: number;
    }

    export interface UploadImagePostMutationRequest {
        /** @description Additional data to pass to server */
        additionalMetadata?: string;
        /** @description file to upload */
        file?: string;
    }

    /** successful operation */
    export type UploadImagePostMutationResponse = ApiResponse;
    export type UploadImagePostError = unknown;
    export type CreateError = unknown;
    export type CreateMutationResponse = unknown;
    export type UpdateError = unknown;
    export type UpdateMutationResponse = unknown;

    /** queryParams */
    export interface FindByStatusGetQueryParams {
        status: Array<string>;
    }

    /** successful operation */
    export type FindByStatusGetQueryResponse = Pet[];
    export type FindByStatusGetError = unknown;

    /** queryParams */
    export interface FindByTagsGetQueryParams {
        tags: Array<string>;
    }

    /** successful operation */
    export type FindByTagsGetQueryResponse = Pet[];
    export type FindByTagsGetError = unknown;

    /** pathParams */
    export interface FindByPetIdPathParams {
        petId: number;
    }

    /** successful operation */
    export type FindByPetIdQueryResponse = Pet;
    export type FindByPetIdError = unknown;

    /** pathParams */
    export interface PetIdPostPathParams {
        petId: number;
    }

    export interface PetIdPostMutationRequest {
        /** @description Updated name of the pet */
        name?: string;
        /** @description Updated status of the pet */
        status?: string;
    }

    export type PetIdPostError = unknown;
    export type PetIdPostMutationResponse = unknown;

    /** pathParams */
    export interface DelByPetIdPathParams {
        petId: number;
    }

    export type DelByPetIdError = unknown;
    export type DelByPetIdMutationResponse = unknown;
}

import type { Order } from "./typeModels";

/**
 * @tag store
 * @description Access to Petstore orders
 * @UUID type-store
 */
export namespace Store {
    export type CreateMutationRequest = Order;
    /** successful operation */
    export type CreateMutationResponse = Order;
    export type CreateError = unknown;

    /** pathParams */
    export interface FindByOrderIdPathParams {
        orderId: number;
    }

    /** successful operation */
    export type FindByOrderIdQueryResponse = Order;
    export type FindByOrderIdError = unknown;

    /** pathParams */
    export interface DelByOrderIdPathParams {
        orderId: number;
    }

    export type DelByOrderIdError = unknown;
    export type DelByOrderIdMutationResponse = unknown;

    /** successful operation */
    export interface InventoryGetQueryResponse {
        [key: string]: unknown;
    }

    export type InventoryGetError = unknown;
}

import type { User } from "./typeModels";

/**
 * @tag user
 * @description Operations about user
 * @UUID type-user
 */
export namespace User {
    export type CreateWithArrayPostError = unknown;
    export type CreateWithArrayPostMutationResponse = unknown;
    export type CreateWithListPostError = unknown;
    export type CreateWithListPostMutationResponse = unknown;

    /** pathParams */
    export interface FindByUsernamePathParams {
        username: string;
    }

    /** successful operation */
    export type FindByUsernameQueryResponse = User;
    export type FindByUsernameError = unknown;

    /** pathParams */
    export interface UsernamePutPathParams {
        username: string;
    }

    export type UsernamePutMutationRequest = User;
    export type UsernamePutError = unknown;
    export type UsernamePutMutationResponse = unknown;

    /** pathParams */
    export interface DelByUsernamePathParams {
        username: string;
    }

    export type DelByUsernameError = unknown;
    export type DelByUsernameMutationResponse = unknown;

    /** queryParams */
    export interface LoginGetQueryParams {
        username: string;
        password: string;
    }

    /** successful operation */
    export type LoginGetQueryResponse = string;
    export type LoginGetError = unknown;
    export type LogoutGetError = unknown;
    export type LogoutGetQueryResponse = unknown;
    export type CreateMutationRequest = User;
    export type CreateError = unknown;
    export type CreateMutationResponse = unknown;
}

import type { Testdata } from "./Testdata";

/** @UUID type-Test32145 */
export interface Test32145 {
    /** @description description */
    code?: number;
    /** @UUID type-Testdata */
    meta?: Testdata;
}

/** @UUID type-Test3214 */
export interface Test3214 {
    /** @description 默认值 */
    columnDefault?: string;
    /** @description description */
    columnLength?: number;
    /** @description description */
    columnName?: string;
    /** @description description */
    columnRemark?: string;
    /** @description description */
    columnScale?: number;
    /** @description description */
    columnType?: string;
    /** @description description */
    delFlag?: boolean;
    formId?: number;
    /** @description id */
    id?: number;
    /** @description description */
    notNull?: boolean;
}

/** @UUID type-Testdata */
export interface Testdata {
    /** @description description */
    number?: number;
    /** @description description */
    numberOfElements?: number;
    /** @description description */
    totalElements?: number;
    /** @description description */
    totalPages?: number;
}

/** @UUID type-Sort */
export interface Sort {
    empty?: boolean;
    sorted?: boolean;
    unsorted?: boolean;
}

import type { Test3214 } from "./Test3214";

/** @UUID type-TestDto3 */
export interface TestDto3 {
    test3214?: Test3214[];
}

import type { TestDto3 } from "./TestDto3";
import type { Testdata } from "./Testdata";

/** @UUID type-TestDto2 */
export interface TestDto2 {
    /** @UUID type-TestDto3 */
    content?: TestDto3;
    /** @UUID type-Testdata */
    meta?: Testdata;
}

/** @UUID type-Test321 */
export interface Test321 {
    id?: number;
}

import type { Test321 } from "./Test321";

/** @UUID type-TestDto */
export interface TestDto {
    /** @description test321 */
    test321?: Test321[];
}

import type { ApiResponse } from "./ApiResponse";

/** @UUID type-RefList */
export interface RefList {
    /** @UUID type-ApiResponse */
    content?: ApiResponse;
    /** @description string */
    string?: string;
    name?: string;
    /** @description title */
    title?: string;
    /** @description type */
    type?: string;
}

/** @UUID type-ApiResponse */
export interface ApiResponse {
    code?: number;
    type?: string;
    message?: string;
}

/** @UUID type-Category */
export interface Category {
    id?: number;
    name?: string;
}

import type { Category } from "./Category";
import type { Tag } from "./Tag";

/** @UUID type-Pet */
export interface Pet {
    id?: number;
    /** @UUID type-Category */
    category?: Category;
    name: string;
    photoUrls: Array<string>;
    tags?: Tag[];
    /** @description pet status in the store */
    status?: string;
}

/** @UUID type-Tag */
export interface Tag {
    id?: number;
    name?: string;
}

/** @UUID type-Order */
export interface Order {
    id?: number;
    petId?: number;
    quantity?: number;
    shipDate?: string;
    /** @description Order Status */
    status?: string;
    complete?: boolean;
}

/** @UUID type-User */
export interface User {
    id?: number;
    username?: string;
    firstName?: string;
    lastName?: string;
    email?: string;
    password?: string;
    phone?: string;
    /** @description User Status */
    userStatus?: number;
}

/** description */
export const ColumnTypeLabel = {
    Char: '',
    Char2: '',
    Char3: ''
};
/** pet status in the store */
export const StatusLabel = {
    Available: '',
    Pending: '',
    Sold: ''
};
/** Order Status */
export const Status1Label = {
    Placed: '',
    Approved: '',
    Delivered: ''
};

/** description */
const enum ColumnType {
    Char = "CHAR",
    Char2 = "CHAR2",
    Char3 = "CHAR3"
}

/** pet status in the store */
const enum Status {
    Available = "available",
    Pending = "pending",
    Sold = "sold"
}

/** Order Status */
const enum Status1 {
    Placed = "placed",
    Approved = "approved",
    Delivered = "delivered"
}

/** description */
export const columnTypeOption = [{
    label: ColumnTypeLabel.Char,
    value: ColumnType.Char
}, {
    label: ColumnTypeLabel.Char2,
    value: ColumnType.Char2
}, {
    label: ColumnTypeLabel.Char3,
    value: ColumnType.Char3
}];
/** pet status in the store */
export const statusOption = [{
    label: StatusLabel.Available,
    value: Status.Available
}, {
    label: StatusLabel.Pending,
    value: Status.Pending
}, {
    label: StatusLabel.Sold,
    value: Status.Sold
}];
/** Order Status */
export const status1Option = [{
    label: Status1Label.Placed,
    value: Status1.Placed
}, {
    label: Status1Label.Approved,
    value: Status1.Approved
}, {
    label: Status1Label.Delivered,
    value: Status1.Delivered
}];

export * from "./enum";
export * from "./Test32145";
export * from "./Test3214";
export * from "./Testdata";
export * from "./Sort";
export * from "./TestDto3";
export * from "./TestDto2";
export * from "./Test321";
export * from "./TestDto";
export * from "./RefList";
export * from "./ApiResponse";
export * from "./Category";
export * from "./Pet";
export * from "./Tag";
export * from "./Order";
export * from "./User";
"
`;

exports[`typerGenerator > compare TypeGenerator getFullText 1`] = `
"import type { TestDto2, Test32145, ApiResponse, Pet } from "./typeModels";

/**
 * @tag pet
 * @description Everything about your Pets
 * @UUID type-pet
 */
export namespace NewPet {
    /** OK */
    export type TestPostMutationResponse = TestDto2;
    export type TestPostError = unknown;
    /** OK */
    export type TestPutMutationResponse = TestDto2;
    export type TestPutError = unknown;
    export type DelByTestMutationRequest = Array<number>;
    /** OK */
    export type DelByTestMutationResponse = Test32145;
    export type DelByTestError = unknown;

    /** queryParams */
    export interface TestIdGetQueryParams {
        fields?: Array<string>;
        page: number;
        size: number;
    }

    /** pathParams */
    export interface TestIdGetPathParams {
        testId?: number;
        testId2?: string;
    }

    /** OK */
    export type TestIdGetQueryResponse = TestDto2;
    export type TestIdGetError = unknown;

    /** pathParams */
    export interface UploadImagePostPathParams {
        petId: number;
    }

    export interface UploadImagePostMutationRequest {
        /** @description Additional data to pass to server */
        additionalMetadata?: string;
        /** @description file to upload */
        file?: string;
    }

    /** successful operation */
    export type UploadImagePostMutationResponse = ApiResponse;
    export type UploadImagePostError = unknown;
    export type CreateError = unknown;
    export type CreateMutationResponse = unknown;
    export type UpdateError = unknown;
    export type UpdateMutationResponse = unknown;

    /** queryParams */
    export interface FindByStatusGetQueryParams {
        status: Array<string>;
    }

    /** successful operation */
    export type FindByStatusGetQueryResponse = Pet[];
    export type FindByStatusGetError = unknown;

    /** queryParams */
    export interface FindByTagsGetQueryParams {
        tags: Array<string>;
    }

    /** successful operation */
    export type FindByTagsGetQueryResponse = Pet[];
    export type FindByTagsGetError = unknown;

    /** pathParams */
    export interface FindByPetIdPathParams {
        petId: number;
    }

    /** successful operation */
    export type FindByPetIdQueryResponse = Pet;
    export type FindByPetIdError = unknown;

    /** pathParams */
    export interface PetIdPostPathParams {
        petId: number;
    }

    export interface PetIdPostMutationRequest {
        /** @description Updated name of the pet */
        name?: string;
        /** @description Updated status of the pet */
        status?: string;
    }

    export type PetIdPostError = unknown;
    export type PetIdPostMutationResponse = unknown;

    /** pathParams */
    export interface DelByPetIdPathParams {
        petId: number;
    }

    export type DelByPetIdError = unknown;
    export type DelByPetIdMutationResponse = unknown;
}

import type { Order } from "./typeModels";

/**
 * @tag store
 * @description Access to Petstore orders
 * @UUID type-store
 */
export namespace Store {
    export type CreateMutationRequest = Order;
    /** successful operation */
    export type CreateMutationResponse = Order;
    export type CreateError = unknown;

    /** pathParams */
    export interface FindByOrderIdPathParams {
        orderId: number;
    }

    /** successful operation */
    export type FindByOrderIdQueryResponse = Order;
    export type FindByOrderIdError = unknown;

    /** pathParams */
    export interface DelByOrderIdPathParams {
        orderId: number;
    }

    export type DelByOrderIdError = unknown;
    export type DelByOrderIdMutationResponse = unknown;

    /** successful operation */
    export interface InventoryGetQueryResponse {
        [key: string]: unknown;
    }

    export type InventoryGetError = unknown;
}

import type { User } from "./typeModels";

/**
 * @tag user
 * @description Operations about user
 * @UUID type-user
 */
export namespace User {
    export type CreateWithArrayPostError = unknown;
    export type CreateWithArrayPostMutationResponse = unknown;
    export type CreateWithListPostError = unknown;
    export type CreateWithListPostMutationResponse = unknown;

    /** pathParams */
    export interface FindByUsernamePathParams {
        username: string;
    }

    /** successful operation */
    export type FindByUsernameQueryResponse = User;
    export type FindByUsernameError = unknown;

    /** pathParams */
    export interface UsernamePutPathParams {
        username: string;
    }

    export type UsernamePutMutationRequest = User;
    export type UsernamePutError = unknown;
    export type UsernamePutMutationResponse = unknown;

    /** pathParams */
    export interface DelByUsernamePathParams {
        username: string;
    }

    export type DelByUsernameError = unknown;
    export type DelByUsernameMutationResponse = unknown;

    /** queryParams */
    export interface LoginGetQueryParams {
        username: string;
        password: string;
    }

    /** successful operation */
    export type LoginGetQueryResponse = string;
    export type LoginGetError = unknown;
    export type LogoutGetError = unknown;
    export type LogoutGetQueryResponse = unknown;
    export type CreateMutationRequest = User;
    export type CreateError = unknown;
    export type CreateMutationResponse = unknown;
}

import type { Testdata } from "./Testdata";

/** @UUID type-Test32145 */
export interface Test32145 {
    /** @description description */
    code?: number;
    /** @UUID type-Testdata */
    meta?: Testdata;
}

/** @UUID type-Test3214 */
export interface Test3214 {
    /** @description 默认值 */
    columnDefault?: string;
    /** @description description */
    columnLength?: number;
    /** @description description */
    columnName?: string;
    /** @description description */
    columnRemark?: string;
    /** @description description */
    columnScale?: number;
    /** @description description */
    columnType?: string;
    /** @description description */
    delFlag?: boolean;
    formId?: number;
    /** @description id */
    id?: number;
    /** @description description */
    notNull?: boolean;
}

/** @UUID type-Testdata */
export interface Testdata {
    /** @description description */
    number?: number;
    /** @description description */
    numberOfElements?: number;
    /** @description description */
    totalElements?: number;
    /** @description description */
    totalPages?: number;
}

/** @UUID type-Sort */
export interface Sort {
    empty?: boolean;
    sorted?: boolean;
    unsorted?: boolean;
}

import type { Test3214 } from "./Test3214";

/** @UUID type-TestDto3 */
export interface TestDto3 {
    test3214?: Test3214[];
}

import type { TestDto3 } from "./TestDto3";
import type { Testdata } from "./Testdata";

/** @UUID type-TestDto2 */
export interface TestDto2 {
    /** @UUID type-TestDto3 */
    content?: TestDto3;
    /** @UUID type-Testdata */
    meta?: Testdata;
}

/** @UUID type-Test321 */
export interface Test321 {
    id?: number;
}

import type { Test321 } from "./Test321";

/** @UUID type-TestDto */
export interface NewTestDto {
    /** @description test321 */
    test321?: Test321[];
}

import type { ApiResponse } from "./ApiResponse";

/** @UUID type-RefList */
export interface RefList {
    /** @UUID type-ApiResponse */
    content?: ApiResponse;
    /** @description string */
    string?: string;
    name?: string;
    /** @description title */
    title?: string;
    /** @description type */
    type?: string;
}

/** @UUID type-ApiResponse */
export interface ApiResponse {
    code?: number;
    type?: string;
    message?: string;
}

/** @UUID type-Category */
export interface NewCategory {
    id?: number;
    name?: string;
}

import type { NewCategory } from "./NewCategory";
import type { Tag } from "./Tag";

/** @UUID type-Pet */
export interface Pet {
    id?: number;
    /** @UUID type-Category */
    category?: NewCategory;
    name: string;
    photoUrls: Array<string>;
    tags?: Tag[];
    /** @description pet status in the store */
    status?: string;
}

/** @UUID type-Tag */
export interface Tag {
    id?: number;
    name?: string;
}

/** @UUID type-Order */
export interface Order {
    id?: number;
    petId?: number;
    quantity?: number;
    shipDate?: string;
    /** @description Order Status */
    status?: string;
    complete?: boolean;
}

/** @UUID type-User */
export interface User {
    id?: number;
    username?: string;
    firstName?: string;
    lastName?: string;
    email?: string;
    password?: string;
    phone?: string;
    /** @description User Status */
    userStatus?: number;
}

/** description */
export const ColumnTypeLabel = {
    Char: '',
    Char2: '',
    Char3: ''
};
/** pet status in the store */
export const StatusLabel = {
    Available: '',
    Pending: '',
    Sold: ''
};
/** Order Status */
export const Status1Label = {
    Placed: '',
    Approved: '',
    Delivered: ''
};

/** description */
const enum ColumnType {
    Char = "CHAR",
    Char2 = "CHAR2",
    Char3 = "CHAR3"
}

/** pet status in the store */
const enum Status {
    Available = "available",
    Pending = "pending",
    Sold = "sold"
}

/** Order Status */
const enum Status1 {
    Placed = "placed",
    Approved = "approved",
    Delivered = "delivered"
}

/** description */
export const columnTypeOption = [{
    label: ColumnTypeLabel.Char,
    value: ColumnType.Char
}, {
    label: ColumnTypeLabel.Char2,
    value: ColumnType.Char2
}, {
    label: ColumnTypeLabel.Char3,
    value: ColumnType.Char3
}];
/** pet status in the store */
export const statusOption = [{
    label: StatusLabel.Available,
    value: Status.Available
}, {
    label: StatusLabel.Pending,
    value: Status.Pending
}, {
    label: StatusLabel.Sold,
    value: Status.Sold
}];
/** Order Status */
export const status1Option = [{
    label: Status1Label.Placed,
    value: Status1.Placed
}, {
    label: Status1Label.Approved,
    value: Status1.Approved
}, {
    label: Status1Label.Delivered,
    value: Status1.Delivered
}];

export * from "./enum";
export * from "./Test32145";
export * from "./Test3214";
export * from "./Testdata";
export * from "./Sort";
export * from "./TestDto3";
export * from "./TestDto2";
export * from "./Test321";
export * from "./NewTestDto";
export * from "./RefList";
export * from "./ApiResponse";
export * from "./NewCategory";
export * from "./Pet";
export * from "./Tag";
export * from "./Order";
export * from "./User";
"
`;
